docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+      return new SomeService();
+    });
+  }
+
+  ExamplePlugin.prototype = Object.create(Plugin.prototype);
+
+  return ExamplePlugin;
+});
+```
+
+Using this plugin then looks like:
+
+```
+define([
+  'mct',
+  './ExamplePlugin',
+  './OtherPlugin'
+], function (mct, ExamplePlugin, OtherPlugin) {
+  var MCT = mct.MCT;
+
+  function MyApplication() {
+    MCT.call(this);
+
+    this.examplePlugin = new ExamplePlugin(this.core);
+    this.otherPlugin = new OtherPlugin(this.examplePlugin);
+
+    this.install(this.examplePlugin);
+    this.install(this.otherPlugin);
+  }
+
+  MyApplication.prototype = Object.create(MCT.prototype);
+
+  return MyApplication;
+});
+```
+
+## Using Extension Points
+
+The services and extensions exposed by providers are retrieved via
+`get` calls to those providers. These calls are expected to occur
+during registration of other extensions, _or_ in the `start` phase
+of the plugin lifecycle.
+
+There are effectively four distinct stages for a plugin:
+
+* __Pre-initialization__. This is the plugin immediately after its
+  constructor call. Any extension points exposed by the plugin are
+  expected to be defined during this stage.
+* __Initialization__. Triggered by calling `initialize`; invokes
+
+For example:
+
+```
+define(['mct', './SomeAction'], function (mct, SomeAction) {
+  var Plugin = mct.Plugin;
+
+  function MyPlugin(core, notificationPlugin) {
+    var notificationServiceProvider =
+      notificationPlugin.notificationServiceProvider;
+
+    Plugin.call(this, function initialize() {
+      // During this stage, extensions may be installed and other
+      // general plugin configuration should occur.
+      // Calls to get should be avoided at this stage, as
+      // providers may not be fully configured.
+      core.actionRegistry.register(function () {
+        // In factory functions, however, get calls are expected;
+        // this is when dependencies actually get injected.
+        // Calls which register/configure extensions should be
+        // avoided at this point. 
+        return new SomeAction(notificationServiceProvider.get());
+      });
+    }, function start() {
+      // Any behavior that should occur when the application starts.
+      // All providers should be fully-configured at this point; `get`
+      // calls may be issued freely at this point, and no more
+      // registration should occur. This stage is not useful to most
+      // plugins and this argument would typically be omitted.
+
+      notifications.notificationServiceProvider.get()
+        .notify("Hello world!");
+    });
+
+    // Code in the constructor is run when the plugin is instantiated;
+    // any extension points exposed by the plugin should be declared
+    // here, typically as providers.
+    this.someRegistry = new mct.Registry();
+  }
+
+  MyPlugin.prototype = Object.create(Plugin.prototype);
+
+  return MyPlugin;
+
+});
+```
+
+## Extending Extensions
```

Comment:
Decorating is just a nice word for monkeypatching.  If we're doing it, we should probably be refactoring something.  But I'm not totally against it, just you know, call it what it is.